### DISCUSSION TITLE
The JanusGraph does not support working with OpenSearch client to work with OpenSearch server

### DISCUSSION THREAD

**Discussion by luxuabc** [JanusGraph/janusgraph#3541] [Q&A]
Hi team,
We use OpenSearch 1.3 server as our backend. we enabled the compatible mode in OpenSearch, so that works fine with JanusGraph. But now there is security issue of elasticsearch client, with the license of elasticsearch changed, we can't upgrade elasticsearch client to higher version above 7.10.2. So we need to change to use opensearch client, but we don't find the package support OpenSearch client to replace the org.janusgraph.janusgraph-es package. Do you have plan to support this?

**Comment by porunov** [comment]
I'm not aware of any work going on regarding adding OpenSearch support to JanusGraph, but JanusGraph is a community driven project. Thus, if you or anyone else want's to add support for OpenSearch then you more than welcome to add that support.

The way I would do that is to reuse `index.[X].elasticsearch.interface` configuration.
Right now `org.janusgraph.diskstorage.es.ElasticSearchSetup` has only one enum value `REST_CLIENT` but you could add `OPEN_SEARCH_CLIENT` and implement `org.janusgraph.diskstorage.es.ElasticSearchClient` to use OpenSearch client instead of ElasticSearch client. That way you would be able to use `index.[X].elasticsearch.interface` to specify OpenSearch client and not elasticsearch client.
In such case you could simply exclude ElasticSearch client from your code and use only OpenSearch implementation.
As a next step refactoring (not necessary) I would actually remove `org.janusgraph.diskstorage.es.ElasticSearchSetup` and replace that enum with simple class path configuration (i.e. to specify class pass of `ElasticSearchClient` implementation instead of enum value for `index.[X].elasticsearch.interface`).
